crates/accelerate/src/high_level_synthesis.rs

```diff
@@ -171,82 +213,767 @@ impl QubitTracker {
+            if target.num_qubits.is_some() {
+                if borrowed_data.use_qubit_indices {
+                    let physical_qubits = qubits.iter().map(|q| PhysicalQubit(q.0)).collect();
+                    target.instruction_supported(name, Some(&physical_qubits))
+                } else {
+                    target.instruction_supported(name, None)
+                }
+            } else {
+                borrowed_data.device_insts.contains(name)
+            }
+        }
+        None => borrowed_data.device_insts.contains(name),
     }
+}
 
-    /// Returns the global-to-local mapping
-    fn to_local_mapping(&self) -> HashMap<usize, usize> {
-        HashMap::from_iter(
-            self.local_to_global
-                .iter()
-                .enumerate()
-                .map(|(i, j)| (*j, i)),
-        )
+/// Check whether an operation does not need to be synthesized.
+fn definitely_skip_op(
+    py: Python,
+    data: &Bound<HighLevelSynthesisData>,
+    op: &PackedOperation,
+    qubits: &[Qubit],
+) -> bool {
+    let borrowed_data: PyRef<'_, HighLevelSynthesisData> = data.borrow();
+
+    if qubits.len() < borrowed_data.min_qubits {
+        return true;
+    }
+
+    if op.directive() {
+        return true;
+    }
+
+    if op.control_flow() {
+        return false;
+    }
+
+    // If the operation is natively supported, we can skip it.
+    if instruction_supported(py, data, op.name(), qubits) {
+        return true;
+    }
+
+    // If there are available plugins for this operation, we should try them
+    // before checking the equivalence library.
+    if borrowed_data.hls_op_names.iter().any(|s| s == op.name()) {
+        return false;
+    }
+
+    if let Some(equiv_lib) = &borrowed_data.equivalence_library {
+        if equiv_lib.borrow(py).has_entry(op) {
+            return true;
+        }
+    }
+
+    false
+}
+
+/// Recursively synthesizes a circuit. This circuit is either the original circuit,
+/// the definition circuit for one of the gates, or a circuit returned by a plugin.
+///
+/// The input to this function is the circuit to be synthesized and the global
+/// qubits over which it is defined.
+///
+/// The output is the synthesized circuit and the global qubits over which it is
+/// defined. Note that by using auxiliary qubits, the output circuit may be defined
+/// over more qubits than the input circuit.
+///
+/// The function also updates in-place the qubit tracker, which keeps track of the
+/// state of each global qubits (whether it's clean, dirty, or cannot be used).
+fn run_on_circuitdata(
+    py: Python,
+    input_circuit: &CircuitData,
+    input_qubits: &[usize],
+    data: &Bound<HighLevelSynthesisData>,
+    tracker: &mut QubitTracker,
+) -> PyResult<(CircuitData, Vec<usize>)> {
+    if input_circuit.num_qubits() != input_qubits.len() {
+        return Err(TranspilerError::new_err(
+            "HighLevelSynthesis: the input to 'run_on_circuitdata' is incorrect.",
```

Comment:
It would be nice if we'd raise more explicit errors here (and in other places), especially now that the pass is in Rust. Otherwise it's hard to guess what went wrong 🙂 E.g. something like
```suggestion
           format!(
               "HighLevelSynthesis: number of input qubits ({}) does not match the circuit size ({})",
                input_circuit.num_qubits(),
                input_qubits.len()
            ),
```

Reply:
Here and other places: improved transpiler errors, see [373c4ad](https://github.com/Qiskit/qiskit/pull/13813/commits/373c4ad4ad95a8046b4cf0085405ec4446e9002d).